tests/olpopova/homework/homework09/test_fraction.py

```diff
@@ -0,0 +1,106 @@
+import pytest
+
+from olpopova.homework.homework09.fraction import Fraction
+from olpopova.homework.homework08.warrior import Warrior
+
+
+def test_str_repr():
+    a = Fraction(2, 4)
+    assert repr(a) == "Fraction(1, 2)"
+    assert str(a) == "'1 / 2'"
+    assert "val: {:dec}".format(a) == "val: 0.5"
+    assert "val: {}".format(a) == "val: '1 / 2'"
+
+
+def test_bool_assert():
+    assert Fraction(1, 2)
+    assert not Fraction(0)
+
+
+def test_logic_operators():
+    assert Fraction(2, 4) == Fraction(5, 10)
+    assert not (Fraction(1, 2) != Fraction(1, 2))
+    assert Fraction(1, 2) > Fraction(1, 3)
+    assert Fraction(1, 3) < Fraction(1, 2)
+    assert Fraction(1, 2) >= Fraction(1, 3)
+    assert Fraction(1, 3) <= Fraction(1, 2)
+    assert Fraction(1, 2) <= Fraction(1, 2)
+    assert Fraction(1, 2) >= Fraction(1, 2)
+    assert Fraction(2, 4) == 0.5
+    assert 0.5 == Fraction(2, 4)
+    assert 6 > Fraction(1, 3)
+    assert Fraction(1, 3) < 6.5
+    with pytest.raises(AttributeError):
+        assert 'dfhf' == Fraction(2, 4)
+        assert Fraction(2, 4) == 'dfhf'
+
+
+def test_add():
+    a = Fraction(2, 4)
+    assert Fraction(2, 4) + Fraction(1, 3) == Fraction(5, 6)
+    assert Fraction(1, 2) + 2 == Fraction(5, 2)
+    assert Fraction(1, 2) + 2.5 == Fraction(3, 1)
+    assert 2 + Fraction(1, 2) == Fraction(5, 2)
+    with pytest.raises(AttributeError):
+        assert 'dfhf' + Fraction(2, 4)
+        assert Fraction(2, 4) + 'dfhf'
+        assert Warrior() + Fraction(1, 2)
+
+    # in place add
+    b = a
+    b += 2
+    assert b is not a
+    assert b == Fraction(5, 2)
+
+
+def test_subtraction():
+    a = Fraction(2, 4)
+    assert Fraction(2, 4) - Fraction(1, 3) == Fraction(1, 6)
+    assert Fraction(5, 2) - 2 == Fraction(1, 2)
+    assert 2 - Fraction(3, 2) == Fraction(1, 2)
+    assert 3.5 - Fraction(3, 2) == Fraction(2, 1)
+    with pytest.raises(AttributeError):
+        assert Fraction(5, 2) - 'dfhf'
+
+    # in place sub
+    b = a
+    b -= 2
+    assert b is not a
+    assert b == Fraction(-3, 2)
+
+
+def multiplication():
+    a = Fraction(2, 4)
+    assert Fraction(1, 3) * 3 == 1
+    assert Fraction(2, 3) * Fraction(12, 4) == 2
+    assert 2 * Fraction(3, 2) == 3
+    with pytest.raises(AttributeError):
+        assert Fraction(5, 2) * 'dfhf'
+
+    # in place mul
+    b = a
+    b *= 2
+    assert b is not a
+    assert b == 1
+
+
+def test_division():
+    a = Fraction(2, 4)
+    assert Fraction(2, 3) / Fraction(2, 3) == 1
+    assert Fraction(2, 3) / Fraction(4, 3) == Fraction(1, 2)
+    assert 2 / Fraction(4, 3) == Fraction(3, 2)
+    with pytest.raises(AttributeError):
+        assert Fraction(5, 2) / 'dfhf'
```

Comment:
good check , but I assume your error is missleading at the moment, and states that you can not divide by NotImplemented object